Add declarative audit device creation mechanism (#1700)
openbao/openbao: website/content/docs/rfcs/config-audit-devices.mdx

---
sidebar_label: Audit Device Configuration
description: |-
  An OpenBao RFC for defining audit devices in the server configuration
  file through the complete audit device life cycle.
---

# Audit Device Configuration

## Summary

After [Vault HCSEC-2025-14 / CVE-2025-6000](https://discuss.hashicorp.com/t/hcsec-2025-14-privileged-vault-operator-may-execute-code-on-the-underlying-host/76033)
and [OpenBao CVE-2025-54997](https://github.com/openbao/openbao/security/advisories/GHSA-xp75-r577-cvhp),
it became apparent that _any_ API-driven audit device creation is unsafe.
We remediate this fully by allowing full management of audit devices from
the configuration file, throughout the life cycle of the device.

## Problem Statement

API-driven audit device creation is fundamentally unsafe. Even without control

> over `prefix`, audit devices such as `file` allow writing to arbitrary files

Reviewer: ```suggestion
over `prefix`, audit devices of type `file` allow writing to arbitrary files
```

> and `socket` allow writing to arbitrary Unix or network sockets. These

Reviewer: ```suggestion
and type `socket` allows writing to arbitrary Unix or network sockets. These
```

fix for the fundamental risks associated with this.

In a design similar to the [declarative self-initialization](/docs/rfcs/self-init),
we opt to define audit devices in the configuration file. We need to make
sure we handle the following scenarios:

1. Creating a new audit device
2. Preventing duplicate audit devices (across API- and config-based creation)
3. Deleting an existing audit device
4. Preventing modification to an existing audit device.

## User-facing Description

We introduce a new repeatable configuration stanza, `audit "<type>" "<path>"`
which configures a new audit device:

```hcl
audit "file" "to-stdout" {
  description = "Default audit device"
  options = {
    file_path = "/dev/stdout"
    log_raw = "false"
  }
}
```

This can be added or removed and will be reprocessed on server restarts and
when `SIGHUP` is sent to the server. It will be ignored on standby nodes.

## Technical Description

This change is largely self-contained to the audit system and server
configuration parsing, with the exception of two necessary hooks:

1. The standard unseal strategy now updates audit devices based on the
   configuration.

> 2. The reload handlers triggered on `SIGHUP` now will reload the seal as
>    well.

Author: Ah audit devices.

```suggestion
2. The reload handlers triggered on `SIGHUP` now will reload the audit devices as
   well.
```
